ui: Create HistogramWindowDuration config value

mrtracy opened the pull request:
Creates a new configuration value "HistogramWindowDuration", which specifies the
approximate duration for which individual samples are retained in our sliding
window histograms.

Previously, histogram windows were initialized using the MetricSampleWindow
value. However, this resulted in #12998; this is caused because the histogram
window is rotated when querying a histogram, not when recording a
sample. In many cases, the timing is such that all samples recorded are
discarded right before the histogram is queried, resulting in a zero value.

Because of this, and additionally issue #12835, we are plumbing a separate,
longer histogram window duration through to all of these metrics. This includes
all existing histograms that were using MetricSampleWindow, and a few histograms
which were using constant values of 1 minute.

An alternative fix for #12998 would be to "tick" histograms, and thus possibly
cause them to rotate their sliding window, whenever a sample is recorded. That
works more generally, but can be computationally expensive by taking a lot of
additional timestamps. By making the histogram window somewhat longer than the
metrics sample window, this issue is alleviated without the additional overhead.

Fixes #12835
Fixes #12998

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/13106)
<!-- Reviewable:end -->

petermattis commented:
Review status: 0 of 15 files reviewed at latest revision, 2 unresolved discussions, all commit checks successful.

---

*[pkg/server/config.go, line 61 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/13106#-KbD2yz_ame0kRQq0A1y:-KbD2yz_ame0kRQq0A1z:b-yjgq7) ([raw file](https://github.com/cockroachdb/cockroach/blob/688fefbfdb5ab1286edb056c767e535825d8d5ac/pkg/server/config.go#L61)):*
> ```Go
> 	defaultScanMaxIdleTime          = 200 * time.Millisecond
> 	defaultMetricsSampleInterval    = 10 * time.Second
> 	defaultHistogramWindowInterval  = 1 * time.Minute
> ```

Should `defaultHistogramWindowInterval` be set in relation to `defaultMetricsSampleInterval`? Something like: `defaultHistogramWindowInterval = defaultMetricsSampleInterval * 6`. And perhaps this constant shouldn't exist at all but you should initialize `Config.HistogramWindowInterval` from `Config.MetricsSampleInterval`.

I don't have an intuitive sense of what this value should be set to. Why 1m? Why not 15s? Or 20s? By setting this to `6*defaultMetricsSampleInterval`, won't the histogram clear just before 1 out of 6 samples?

---

*[pkg/server/config.go, line 151 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/13106#-KbD3JWnkxFooS233XJk:-KbD3JWnkxFooS233XJl:bihsl7r) ([raw file](https://github.com/cockroachdb/cockroach/blob/688fefbfdb5ab1286edb056c767e535825d8d5ac/pkg/server/config.go#L151)):*
> ```Go
> 	// individual samples are retained for metric histograms.
> 	// Environment Variable: COCKROACH_HISTOGRAM_WINDOW_INTERVAL
> 	HistogramWindowInterval time.Duration
> ```

You've got a nice explanatory comment in the commit message that deserves to be included here. It isn't obvious that setting this to the same value as `COCKROACH_METRICS_SAMPLE_INTERVAL` will cause bad things to happen.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/13106)*
<!-- Sent from Reviewable.io -->

mrtracy commented:
Review status: 0 of 15 files reviewed at latest revision, 2 unresolved discussions, all commit checks successful.

---

*[pkg/server/config.go, line 61 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/13106#-KbD2yz_ame0kRQq0A1y:-KbHKYTBXwzwpFP9sLSU:b-8qgs3i) ([raw file](https://github.com/cockroachdb/cockroach/blob/688fefbfdb5ab1286edb056c767e535825d8d5ac/pkg/server/config.go#L61)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

Should `defaultHistogramWindowInterval` be set in relation to `defaultMetricsSampleInterval`? Something like: `defaultHistogramWindowInterval = defaultMetricsSampleInterval * 6`. And perhaps this constant shouldn't exist at all but you should initialize `Config.HistogramWindowInterval` from `Config.MetricsSampleInterval`.

I don't have an intuitive sense of what this value should be set to. Why 1m? Why not 15s? Or 20s? By setting this to `6*defaultMetricsSampleInterval`, won't the histogram clear just before 1 out of 6 samples?
</blockquote></details>

In terms of computing the counter - I would be fine with setting it based on a constant multiple of the sample interval.  I could even just make it a function of Config rather than a value.

In terms of its value - the issue is, that this value shouldn't exist at all.  We are recording histograms in a very incorrect way, but fixing that would require very significant additions to our time series system.

Some time series systems allow the recording of every individual sample in a histogram; at that point, you can retrieve the histogram of samples for any arbitrary period of time.  However, you have to provide storage for every sample; that is almost certainly prohibitive.

Other time series DBs (such as prometheus) regularly record the cumulative bucket counts for the histogram over time.  Essentially, rather than recording a single time series, they record a time series for each bucket count. With this information, you can get the differential bucket counts between any two time series points, and thus get the correct histogram for that period of time.  In my opinion, given adequate engineering resources, this is what we should be doing.

What WE are doing is keeping a sliding-window histogram in memory, and periodically recording quantiles based on it.  This is very inexact; samples are discarded in batches, and thus there is a significant variance in the time period of samples between quantiles. We also have to decide on this "sample duration" of the histogram at recording time; and, as illustrated here, it may be different than the recording window, thus leading to possibly unintuitive effects.  It also prevents us from displaying actual histograms, and it even prevents us from computing quantiles for different time periods (i.e. longer time periods); we also can't compute aggregate quantiles accurately across the cluster.

So what this change represents is a very fast fix to a problem that would take *considerably* more effort to fix correctly.   The choice of 1 minute is arbitrary; but any other value would be just as arbitrary.

As for the "clear 1 out of 6 samples", its actually a bit more than that because the sliding histogram only has 2 windows by default; therefore, every time it ticks it clear out about half of the samples it contains.  We could increase this number, but that comes with increased memory costs.

I will record all of this to a github issue for posterity.

---

*[pkg/server/config.go, line 151 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/13106#-KbD3JWnkxFooS233XJk:-KbHOAtuxUeEPapu7Go6:bhgnbgm) ([raw file](https://github.com/cockroachdb/cockroach/blob/688fefbfdb5ab1286edb056c767e535825d8d5ac/pkg/server/config.go#L151)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

You've got a nice explanatory comment in the commit message that deserves to be included here. It isn't obvious that setting this to the same value as `COCKROACH_METRICS_SAMPLE_INTERVAL` will cause bad things to happen.
</blockquote></details>

I think i'll move this to a function and make it a simple multiple of time.Duration.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/13106)*
<!-- Sent from Reviewable.io -->

mrtracy commented:
Review status: 0 of 15 files reviewed at latest revision, 2 unresolved discussions, all commit checks successful.

---

*[pkg/server/config.go, line 61 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/13106#-KbD2yz_ame0kRQq0A1y:-KbHSLQqeVsIsy9VERPO:b-drn5sp) ([raw file](https://github.com/cockroachdb/cockroach/blob/688fefbfdb5ab1286edb056c767e535825d8d5ac/pkg/server/config.go#L61)):*
<details><summary><i>Previously, mrtracy (Matt Tracy) wrote…</i></summary><blockquote>

In terms of computing the counter - I would be fine with setting it based on a constant multiple of the sample interval.  I could even just make it a function of Config rather than a value.

In terms of its value - the issue is, that this value shouldn't exist at all.  We are recording histograms in a very incorrect way, but fixing that would require very significant additions to our time series system.

Some time series systems allow the recording of every individual sample in a histogram; at that point, you can retrieve the histogram of samples for any arbitrary period of time.  However, you have to provide storage for every sample; that is almost certainly prohibitive.

Other time series DBs (such as prometheus) regularly record the cumulative bucket counts for the histogram over time.  Essentially, rather than recording a single time series, they record a time series for each bucket count. With this information, you can get the differential bucket counts between any two time series points, and thus get the correct histogram for that period of time.  In my opinion, given adequate engineering resources, this is what we should be doing.

What WE are doing is keeping a sliding-window histogram in memory, and periodically recording quantiles based on it.  This is very inexact; samples are discarded in batches, and thus there is a significant variance in the time period of samples between quantiles. We also have to decide on this "sample duration" of the histogram at recording time; and, as illustrated here, it may be different than the recording window, thus leading to possibly unintuitive effects.  It also prevents us from displaying actual histograms, and it even prevents us from computing quantiles for different time periods (i.e. longer time periods); we also can't compute aggregate quantiles accurately across the cluster.

So what this change represents is a very fast fix to a problem that would take *considerably* more effort to fix correctly.   The choice of 1 minute is arbitrary; but any other value would be just as arbitrary.

As for the "clear 1 out of 6 samples", its actually a bit more than that because the sliding histogram only has 2 windows by default; therefore, every time it ticks it clear out about half of the samples it contains.  We could increase this number, but that comes with increased memory costs.

I will record all of this to a github issue for posterity.
</blockquote></details>

There was already an issue for this, #7896; quoted most of this comment there to add more information.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/13106)*
<!-- Sent from Reviewable.io -->

petermattis commented:
Review status: 0 of 15 files reviewed at latest revision, 2 unresolved discussions, all commit checks successful.

---

*[pkg/server/config.go, line 61 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/13106#-KbD2yz_ame0kRQq0A1y:-KbKqBgzsQlrShlhHYUJ:bpm99v) ([raw file](https://github.com/cockroachdb/cockroach/blob/688fefbfdb5ab1286edb056c767e535825d8d5ac/pkg/server/config.go#L61)):*
<details><summary><i>Previously, mrtracy (Matt Tracy) wrote…</i></summary><blockquote>

There was already an issue for this, #7896; quoted most of this comment there to add more information.
</blockquote></details>

We export histograms to prometheus as a set of cumulative counts in buckets, though it looks like the bucket sizes can vary across nodes. I wonder if that is fouling up any math prometheus does when summing across nodes. Cc @mberhault 

I have no intuition for how the existing histogram code can possibly give us good latency metrics using the quantile approach. And providing a good latency metric seems like something we have to get right. For our own sanity if for nothing else. I think we should move to a latency metric model of fixed buckets. Currently we record 8 values per histogram. Could we move to a model where we have the following fixed buckets: `<100us`, `<1ms`, `<10ms`, `<100ms`, `<1s`, `<10s`, `<100s`, `<inf`. If this is insufficient resolution, we could move down to doubling the bucket size instead of powers-of-10. Let's discuss in person.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/13106)*
<!-- Sent from Reviewable.io -->

petermattis commented:
Review status: 0 of 15 files reviewed at latest revision, 2 unresolved discussions, all commit checks successful.

---

*[pkg/server/config.go, line 61 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/13106#-KbD2yz_ame0kRQq0A1y:-KbLCU_hjQwnygZ7NVka:b-c7d00g) ([raw file](https://github.com/cockroachdb/cockroach/blob/688fefbfdb5ab1286edb056c767e535825d8d5ac/pkg/server/config.go#L61)):*
> We export histograms to prometheus as a set of cumulative counts in buckets, though it looks like the bucket sizes can vary across nodes. I wonder if that is fouling up any math prometheus does when summing across nodes. Cc @mberhault

Actually, bucket sizes do not vary across nodes. But we configure hdrhistogram to use a shit-ton of buckets. `metric.NewLatency` currently creates a histogram with 483 buckets, 268 of which are for granularities less than 1ms. That's likely overkill. hdrhistogram also enforces a certain amount of precision in its bucketing. If we got rid of that enforcement, we could configure it to have a range from `1ms-10s` using 16 buckets (essentially doubling in size for each bucket).

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/13106)*
<!-- Sent from Reviewable.io -->

Commit:
ui: Compute longer Histogram Windows

Adds a derived configuration value "HistogramWindowDuration" to the server
config, which specifies the approximate duration for which individual samples
are retained in our sliding window histograms.

Previously, histogram windows were initialized using the MetricSampleWindow
value. However, this resulted in #12998; this is caused because the histogram
window is rotated when querying a histogram, not when recording a
sample. In many cases, the timing is such that all samples recorded are
discarded right before the histogram is queried, resulting in a zero value.

Because of this, and additionally issue #12835, we are lengthening the histogram
window duration for all histogram metrics. This includes all existing histograms
that were using MetricSampleWindow, and a few histograms which were using
constant values of 1 minute. The new value is arbitrarily chosen to be six times
the MetricSampleWindow configuration setting; under default settings, that
equates to a window of one minute.

The work here can be considered a temporary fix to alleviate the problems in a
couple of issues; the histogram system is going to be greatly overhauled in
issue #7896, part of which will be the removal of sliding-window histograms.

Fixes #12835
Fixes #12998

mrtracy commented:
Modified this a bit: there is no new configuration value, instead Histogram windows are set to six times the metrics sample interval.  

It is noted in comments and the commit message that this work is a temporary fix until #7896 can be completed. However, I still think this should be checked in; although it has problems, the results should be somewhat more useful than the zeroes we are currently seeing.

petermattis commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Review status: 0 of 13 files reviewed at latest revision, 1 unresolved discussion, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/13106#-:-Kc3UEFYVn5M6Z1RL3X2:bnfp4nl)*
<!-- Sent from Reviewable.io -->
